### DISCUSSION TITLE
Installation on Ubuntu

### DISCUSSION THREAD

**Discussion by helena-intel** [openvinotoolkit/openvino_notebooks#58] [General]
You may need to install some additional libraries on Ubuntu Linux. These steps work on a clean install of Ubuntu Desktop 20.04, and should also work on Ubuntu 18.04 and 20.10, and on Ubuntu Server..

```
sudo apt-get update
sudo apt-get upgrade
sudo apt-get install python3-venv build-essential python3-dev
```

After this, you can follow the instructions in the [README](https://github.com/openvinotoolkit/openvino_notebooks). In step 2, make sure to type `python3 -m venv openvino_env`. `python3` will point to the right Python version. 

On Ubuntu 18.04, `python3-dev` installs the required libraries for the system default version of Python.  On Ubuntu 18 this is Python 3.6,  on Ubuntu 20.04, Python 3.8.  If you also installed other versions of Python, it is recommended to use the full path the to system default Python: `/usr/bin/python3.6 -m venv openvino_env` on Ubuntu 18, `/usr/bin/python3.8 -m venv openvino_env` on Ubuntu 20.

**Comment by raymondlo84** [comment]
Ok, this saved me time again, and for others this is the error message if you are using Python 3.8...
```
Running setup.py install for editdistance ... error
    ERROR: Command errored out with exit status 1:
     command: /home/raymondlo84/openvino_notebooks/openvino_env/bin/python -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-sbd6bhy7/editdistance/setup.py'"'"'; __file__='"'"'/tmp/pip-install-sbd6bhy7/editdistance/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-crokyxvd/install-record.txt --single-version-externally-managed --compile --install-headers /home/raymondlo84/openvino_notebooks/openvino_env/include/site/python3.8/editdistance
         cwd: /tmp/pip-install-sbd6bhy7/editdistance/
    Complete output (16 lines):
    running install
    running build
    running build_py
    creating build
    creating build/lib.linux-x86_64-3.8
    creating build/lib.linux-x86_64-3.8/editdistance
    copying editdistance/__init__.py -> build/lib.linux-x86_64-3.8/editdistance
    copying editdistance/_editdistance.h -> build/lib.linux-x86_64-3.8/editdistance
    copying editdistance/def.h -> build/lib.linux-x86_64-3.8/editdistance
    running build_ext
    building 'editdistance.bycython' extension
    creating build/temp.linux-x86_64-3.8
    creating build/temp.linux-x86_64-3.8/editdistance
    x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -I./editdistance -I/home/raymondlo84/openvino_notebooks/openvino_env/include -I/usr/include/python3.8 -c editdistance/_editdistance.cpp -o build/temp.linux-x86_64-3.8/editdistance/_editdistance.o
    unable to execute 'x86_64-linux-gnu-gcc': No such file or directory
    error: command 'x86_64-linux-gnu-gcc' failed with exit status 1
    ----------------------------------------
ERROR: Command errored out with exit status 1: /home/raymondlo84/openvino_notebooks/openvino_env/bin/python -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-sbd6bhy7/editdistance/setup.py'"'"'; __file__='"'"'/tmp/pip-install-sbd6bhy7/editdistance/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-crokyxvd/install-record.txt --single-version-externally-managed --compile --install-headers /home/raymondlo84/openvino_notebooks/openvino_env/include/site/python3.8/editdistance Check the logs for full command output.
```

**Comment by meng032020** [comment]
Hi, I encountered the error below.
I'm using Ubuntu 18.04, python3.8 in a VirtualBox (if it makes a difference here).
Following the additional steps. I saw the below errors. 
![openvino_error on ubuntu](https://user-images.githubusercontent.com/69293028/118820485-e3e08e00-b8e8-11eb-8224-0dc2c637af68.PNG)

It has the "importing fails" and suggested some workarounds. But I have these errors: 
![error ubuntu openvino2](https://user-images.githubusercontent.com/69293028/118820957-581b3180-b8e9-11eb-80a9-76c04c67ef8c.PNG)

Appreciate any advice that you have. Thanks!

**Reply by raymondlo84** [reply]
Um, I think you have to make sure you are on the openvino_env ? I see abc123? Let's make sure we don't change those substeps?

**Reply by meng032020** [reply]
Hi @raymondlo84 , yes, it is just a name for the folder and the environment. because I had previously done the openvino_env, that's why I am using a different environment for testing. thanks for the advice.

**Reply by meng032020** [reply]
![error ubuntu openvino3](https://user-images.githubusercontent.com/69293028/118908068-41a9c000-b953-11eb-9794-aa4569d7f268.PNG)

hi, using the steps as listed in the guide and the same venv name, I still encounter the error. please help to advise. thanks!

**Reply by raymondlo84** [reply]
I think you have to run something like 

apt-get install python3.8-dev

https://packages.ubuntu.com/search?keywords=python3.8-dev

I assume you have 3.7 and 3.8 installed? Also, I would recommend using python 3.7, and you can basically replace the "python3 -m venv openvino_env" step with the "python3.7 -m venv openvino_env" and follow the rest. 

Having multiple pythons is often pretty confusing! 
-ray

**Reply by helena-intel** [reply]
@meng032020 it appears that you are in a conda environment with Python 3.8. On Ubuntu 18.04, Python 3.6 is the default Python, and installing libpython3-dev will have installed the Python 3.6 libraries. The easiest way to get started on Ubuntu 18.04 is to use Python 3.6. 

I made a wiki page with instructions for conda that use conda envs instead of Python venvs. Could you try if that works for you? You can delete the virtualenvs you made previously.  https://github.com/openvinotoolkit/openvino_notebooks/wiki/Conda

I am sorry you ran into issues! It is tricky to get the instructions working for every scenario on every OS. Thank you for the screenshot, this helps a lot. I will use this to improve the check_install.py script, and by default recommend Python 3.6 on Ubuntu 18. I'll also mention that in the Ubuntu instructions.

**Reply by meng032020** [reply]
Hi @raymondlo84 @helena-intel , thanks for the advice... I'll try them out later when I have access to the system. Thanks!
